hazelcast/src/hazelcast/client/connection/ClientConnectionManagerImpl.cpp

```diff
@@ -0,0 +1,842 @@
+                    connection = firstCallback->get();
+
+                    if (!asOwner) {
+                        return connection;
+                    }
+                    if (connection->isAuthenticatedAsOwner()) {
+                        return connection;
+                    }
+                }
+            }
+
+            void
+            ClientConnectionManagerImpl::authenticate(const Address &target, boost::shared_ptr<Connection> &connection,
+                                                      bool asOwner, boost::shared_ptr<AuthenticationFuture> &future) {
+                const boost::shared_ptr<protocol::Principal> &principal = getPrincipal();
+                std::auto_ptr<protocol::ClientMessage> clientMessage = encodeAuthenticationRequest(asOwner,
+                                                                                                   client.getSerializationService(),
+                                                                                                   principal.get());
+                boost::shared_ptr<spi::impl::ClientInvocation> clientInvocation = spi::impl::ClientInvocation::create(
+                        client, clientMessage, "", connection);
+                boost::shared_ptr<spi::impl::ClientInvocationFuture> invocationFuture = spi::impl::ClientInvocation::invokeUrgent(
+                        clientInvocation);
+                // TODO: let this return a future and pass it to AuthCallback as in Java
+                executionService.schedule(
+                        boost::shared_ptr<util::Runnable>(new TimeoutAuthenticationTask(invocationFuture, *this)),
+                        connectionTimeoutMillis);
+                invocationFuture->andThen(
+                        boost::shared_ptr<impl::ExecutionCallback<boost::shared_ptr<protocol::ClientMessage> > >(
+                                new AuthCallback(connection, asOwner, target, future, *this)));
+            }
+
+            const boost::shared_ptr<protocol::Principal> ClientConnectionManagerImpl::getPrincipal() {
+                return principal;
+            }
+
+            std::auto_ptr<protocol::ClientMessage>
+            ClientConnectionManagerImpl::encodeAuthenticationRequest(bool asOwner,
+                                                                     serialization::pimpl::SerializationService &ss,
+                                                                     const protocol::Principal *principal) {
+                byte serializationVersion = ss.getVersion();
+                const std::string *uuid = NULL;
+                const std::string *ownerUuid = NULL;
+                if (principal != NULL) {
+                    uuid = principal->getUuid();
+                    ownerUuid = principal->getOwnerUuid();
+                }
+                std::auto_ptr<protocol::ClientMessage> clientMessage;
+                const Credentials *credentials = client.getClientConfig().getCredentials();
```

Comment:
creadentials is already set to a class field at the start. No need to get it from config again/

Reply:
Ok, removed.